### PR TITLE
Implement thread local storage

### PR DESCRIPTION
**Pull request by lukileczo**
## Description

- Implements thread local storage on all supported architectures.
## Motivation and Context

JIRA: RTOS-209
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [x] New test added: https://github.com/phoenix-rtos/phoenix-rtos-tests/pull/125.
- [x] Tested by hand on: `ia32-generic-qemu`, `armv7m7-imxrt106x-evk`, `riscv64-generic-qemu`, `armv7a9-zynq7000-qemu`

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work https://github.com/phoenix-rtos/phoenix-rtos-kernel/pull/340, https://github.com/phoenix-rtos/phoenix-rtos-build/pull/127, https://github.com/phoenix-rtos/libphoenix/pull/195.
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: hal/armv7a/cpu.c [hunk 1]
```diff
@@ -19,6 +19,7 @@
 #include "../cpu.h"
 #include "../string.h"
 #include "../spinlock.h"
+#include "../../proc/process.h"
```
Comment (agkaminski) [inline hal/armv7a/cpu.c:22]: top -> down module referencing (i.e. using proc API in HAL) is forbidden. `tls_t` definition should be located in HAL (and named `hal_tls_t`), this will solve this issue.

File: .gitignore [hunk 2]
```diff
@@ -11,3 +11,4 @@
 CMakeLists.txt
 .idea/
 cmake-build-debug/
+.vscode/*
```
Comment (agkaminski) [inline .gitignore:14]: This change is ok, but should be in a separate commit

File: hal/armv7a/cpu.c [hunk 3]
```diff
@@ -199,3 +200,15 @@ char *hal_cpuFeatures(char *features, unsigned int len)
...
 	return features;
 }
+
+
+void hal_cpuTlsSet(tls_t *tls, cpu_context_t *ctx)
+{
+	/* In theory there should be 8-byte thread control block but
+	 * it's stored elsewhere so we need to subtract 8 from the pointer
+	 */
+	u32 ptr = (u32)tls->tls_base - 8;
```
Comment (agkaminski) [inline hal/armv7a/cpu.c:210]: Why is it stored in a different place than expected? This might be covering some hidden bug elsewhere.
Also `ptr_t` should be used in place of `u32`.

File: hal/ia32/cpu.c [hunk 4]
```diff
@@ -23,6 +23,8 @@
 #include "pci.h"
 #include "ia32.h"
 #include "halsyspage.h"
+#include "../../lib/printf.h"
```
Comment (agkaminski) [inline hal/ia32/cpu.c:26]: Debugging left-over, I guess
```suggestion
```

File: hal/ia32/cpu.c [hunk 5]
```diff
@@ -416,15 +417,14 @@ void *_cpu_initCore(void)
...
+	_cpu_gdtInsert(TLS_DESC_IDX + cpu.ncpus, (u32)&cpu.tss[hal_cpuGetID()], sizeof(tss_t), DESCR_TSS);
 
 	cpu.tss[hal_cpuGetID()].ss0 = SEL_KDATA;
 	cpu.tss[hal_cpuGetID()].esp0 = (u32)&cpu.stacks[hal_cpuGetID()][511];
 
 	/* Set task register */
 	__asm__ volatile(
-		"ltr %0; "
-	:: "r" ((u16)((4 + cpu.ncpus) * 8)));
+		"ltr %0; " ::"r"((u16)((TLS_DESC_IDX + cpu.ncpus) * 8)));
```
Comment (agkaminski) [inline hal/ia32/cpu.c:427]: Keep previous formatting

File: hal/ia32/cpu.c [hunk 6]
```diff
@@ -605,3 +606,25 @@ void _hal_cpuInit(void)
 //	hal_cpuDebugGuard(1, 3);
 #endif
 }
+
+
+void hal_cpuTlsSet(tls_t *tls, cpu_context_t *ctx)
+{
+	/* At the end of TLS there must be pointer to itself */
```
Comment (agkaminski) [inline hal/ia32/cpu.c:613]: Coding style: in kernel all variables must be declared before logic.

File: hal/ia32/cpu.c [hunk 7]
```diff
@@ -605,3 +606,25 @@ void _hal_cpuInit(void)
...
+	*(ptr_t *)((ptr_t)tls->tls_base + tls->tdata_sz + tls->tbss_sz) = (ptr_t)tls->tls_base + tls->tdata_sz + tls->tbss_sz;
+
+	ctx->gs = SEL_TLS;
+	u32 base = (u32)((char *)tls->tls_base + tls->tbss_sz + tls->tdata_sz);
+	u32 limit = VADDR_KERNEL;
```
Comment (agkaminski) [inline hal/ia32/cpu.c:618]: These should be of `ptr_t` type

Comment (agkaminski) [reply hal/ia32/cpu.c:618]: I think descriptors should be also, I guess these also scale up to 64 bits on x64, but I'm not sure

File: proc/process.c [hunk 8]
```diff
@@ -409,19 +432,49 @@ int process_load32(vm_map_t *map, vm_object_t *o, offs_t base, void *iehdr)
 			hal_memset(vaddr + filesz, 0, round_page((unsigned long)vaddr + memsz) - (unsigned long)vaddr - filesz);
 		}
 	}
+	shdr = (void *)((char *)ehdr + ehdr->e_shoff);
+	shdr += ehdr->e_shstrndx;
+
+	char *snameTab = (char *)ehdr + shdr->sh_offset;
```
Comment (agkaminski) [inline proc/process.c:438]: Variables should be declared on the beginning of the function

File: proc/process.c [hunk 9]
```diff
@@ -361,18 +364,38 @@ static void process_illegal(unsigned int n, exc_context_t *ctx)
...
+	process->tls.tls_sz = tls_size(&process->tls);
+	process->tls.arm_m_tls = tls->arm_m_tls;
+}
 
 
+#ifndef NOMMU
+
 /* TODO - adding error handling and unmapping of already mapped segments */
-int process_load32(vm_map_t *map, vm_object_t *o, offs_t base, void *iehdr)
+int process_load32(process_t *process, vm_map_t *map, vm_object_t *o, offs_t base, void *iehdr)
```
Comment (agkaminski) [inline proc/process.c:388]: Imho it would be better to pass a pointer to tls here, instead of process. Then we can easily make tls init and assignment code common between load32 and load64

File: proc/process.c [hunk 10]
```diff
@@ -1328,3 +1446,31 @@ int _process_init(vm_map_t *kmap, vm_object_t *kernel)
...
+void tls_init(tls_t *dest, tls_t *source, vm_map_t *map)
+{
+	dest->tdata_sz = source->tdata_sz;
+	dest->tbss_sz = source->tbss_sz;
+	dest->tls_sz = round_page(source->tls_sz);
+	dest->arm_m_tls = source->arm_m_tls;
+
+	dest->tls_base = vm_mmap(map, NULL, NULL, dest->tls_sz, PROT_READ | PROT_WRITE | PROT_USER, NULL, 0, MAP_NONE);
+
+	if (dest->tls_base != NULL) {
```
Comment (agkaminski) [inline proc/process.c:1460]: We should return -ENOMEM if it was indeed NULL - we shouldn't "successfully" spawn a thread that expects TLS without TLS

File: hal/armv7a/cpu.c (hunk 3 above)
Comment (lukileczo) [reply hal/armv7a/cpu.c:210]: Ok, maybe I've misunderstood something, but GCC generates calls for first TLS variable 8 bytes over the address in the register. ARM abi only specifies TLS for Linux.

File: .gitignore (hunk 2 above)
Comment (lukileczo) [reply .gitignore:14]: It is

Comment (agkaminski) [reply .gitignore:14]: Oops, sorry!

File: hal/armv7a/cpu.c [hunk 11]
```diff
@@ -199,3 +200,15 @@ char *hal_cpuFeatures(char *features, unsigned int len)
...
+
+void hal_cpuTlsSet(hal_tls_t *tls, cpu_context_t *ctx)
+{
+	/* In theory there should be 8-byte thread control block but
+	 * it's stored elsewhere so we need to subtract 8 from the pointer
+	 */
+	ptr_t ptr = (ptr_t)tls->tls_base - 8;
+	__asm__ volatile("mov r0, %[value];"
+					 "mcr p15, 0, r0, cr13, cr0, 3;" ::[value] "r"(ptr)
+					 : "r0");
```
Comment (kemonats) [inline hal/armv7a/cpu.c:213]: One `mov` instruction less
```suggestion
	__asm__ volatile("mcr p15, 0, %[value], cr13, cr0, 3;" ::[value] "r"(ptr));
```

File: proc/process.c [hunk 12]
```diff
@@ -361,18 +364,39 @@ static void process_illegal(unsigned int n, exc_context_t *ctx)
 }
 
 
-#ifndef NOMMU
+static void assign_tls(hal_tls_t *process_tls, hal_tls_t *tls, void *tbss_addr)
```
Comment (agkaminski) [inline proc/process.c:367]: Use module prefix, on what -> do what nomenclature and camelCase:
`process_tlsAssign`

File: proc/process.c [hunk 13]
```diff
@@ -1328,3 +1450,37 @@ int _process_init(vm_map_t *kmap, vm_object_t *kernel)
 	hal_exceptionsSetHandler(EXC_UNDEFINED, process_illegal);
 	return EOK;
 }
+
+
+int tls_init(hal_tls_t *dest, hal_tls_t *source, vm_map_t *map)
```
Comment (agkaminski) [inline proc/process.c:1455]:
```suggestion
int process_tlsInit(hal_tls_t *dest, hal_tls_t *source, vm_map_t *map)
```

File: proc/process.c [hunk 14]
```diff
@@ -809,6 +924,13 @@ static void process_exec(thread_t *current, process_spawn_t *spawn)
 		hal_spinlockClear(&spawn->sl, &sc);
 	}
 
+	if (current->process->tls.tls_base != NULL) {
+		err = tls_init(&current->tls, &current->process->tls, current->process->mapp);
+		if (err == EOK) {
+			hal_cpuTlsSet(&current->tls, current->context);
```
Comment (agkaminski) [inline proc/process.c:930]: What will happen if user code that expects TLS to be present gets a process without TLS? I think it will crash or malfunction in a bad way (i.e. variables we expect to be thread safe are not). I think exec should fail if tls_base != NULL and tls_init failed.

File: proc/process.c [hunk 15]
```diff
@@ -1328,3 +1450,37 @@ int _process_init(vm_map_t *kmap, vm_object_t *kernel)
...
+		err = EOK;
+	}
+	else {
+		err = -ENOMEM;
+	}
+	return err;
+}
+
+
+size_t tls_size(hal_tls_t *tls)
```
Comment (agkaminski) [inline proc/process.c:1477]: Used in one place, this function is not needed

File: proc/threads.c [hunk 16]
```diff
@@ -787,68 +791,82 @@ int proc_threadCreate(process_t *process, void (*start)(void *), unsigned int *i
...
-		return -EINVAL;
+	if (priority >= sizeof(threads_common.ready) / sizeof(thread_t *)) {
+		err = -EINVAL;
+	}
 
-	if ((t = vm_kmalloc(sizeof(thread_t))) == NULL)
-		return -ENOMEM;
+	if ((t = vm_kmalloc(sizeof(thread_t))) == NULL) {
+		err = -ENOMEM;
+	}
```
Comment (agkaminski) [inline proc/threads.c:802]: This code refactor is out-of-scope of this PR (and not ideal too, if we fail here there is no need for stack allocation below, also we should extract vm_kmalloc call from if to be MISRA friendly).

File: proc/threads.c [hunk 17]
```diff
@@ -787,68 +791,82 @@ int proc_threadCreate(process_t *process, void (*start)(void *), unsigned int *i
...
+			hal_cpuSetCtxGot(t->context, process->got);
 
-		hal_spinlockClear(&threads_common.spinlock, &sc);
-		hal_cpuSetCtxGot(t->context, process->got);
-	}
+			if (process->tls.tdata_sz != 0 || process->tls.tbss_sz != 0) {
+				err = tls_init(&t->tls, &process->tls, process->mapp);
+				if (err == EOK) {
+					hal_cpuTlsSet(&t->tls, t->context);
+				}
```
Comment (agkaminski) [inline proc/threads.c:854]: We should fail if we cannot provide TLS for a thread that expects it

File: proc/process.c (hunk 14 above)
Comment (lukileczo) [reply proc/process.c:930]: I think the check `if (err < 0) {` lower, in line 934 covers tls_init failure

**Review by agkaminski — APPROVED** [review]
LGTM, let's wait for test in CI

File: hal/armv7a/cpu.c [hunk 18]
```diff
@@ -199,3 +200,14 @@ char *hal_cpuFeatures(char *features, unsigned int len)
...
+
+
+void hal_cpuTlsSet(hal_tls_t *tls, cpu_context_t *ctx)
+{
+	/* In theory there should be 8-byte thread control block but
+	 * it's stored elsewhere so we need to subtract 8 from the pointer
+	 */
+	ptr_t ptr = (ptr_t)tls->tls_base - 8;
+	__asm__ volatile("mcr p15, 0, %[value], cr13, cr0, 3;" ::[value] "r"(ptr)
+					 : "r0");
```
Comment (kemonats) [inline hal/armv7a/cpu.c:212]: The `r0` is not needded in the clobbers list. You are not using the register `r0` explicitly.
```suggestion
	__asm__ volatile("mcr p15, 0, %[value], cr13, cr0, 3;" ::[value] "r"(ptr));
```

Comment (lukileczo) [reply hal/armv7a/cpu.c]: Removed `r0` from clobbers

**Review by agkaminski — APPROVED** [review]
